catalog/app/containers/Bucket/PackageUpdateDialog.js

```diff
@@ -25,7 +26,460 @@ import * as validators from 'utils/validators'
+  },
+  body: {
+    paddingLeft: 20,
+  },
+  bar: {
+    bottom: 0,
+    left: 0,
+    opacity: 0.2,
+    position: 'absolute',
+    top: 24,
+    width: 24,
+    '$active > $head > &, $head:hover > &': {
+      opacity: 0.4,
+    },
+    '&::before': {
+      background: 'currentColor',
+      borderRadius: 2,
+      bottom: 4,
+      content: '""',
+      left: 10,
+      position: 'absolute',
+      top: 4,
+      width: 4,
+    },
+  },
+  emptyDummy: {
+    height: 24,
+  },
+  empty: {
+    ...t.typography.body2,
+    bottom: 0,
+    fontStyle: 'italic',
+    left: 24,
+    lineHeight: '24px',
+    opacity: 0.6,
+    overflow: 'hidden',
+    position: 'absolute',
+    right: 0,
+    top: 24,
+  },
+}))
+
+function Dir({ name, type, children, prefix, dispatch }) {
+  const classes = useDirStyles()
+  // TODO: move state out?
+  const [expanded, setExpanded] = React.useState(true)
+
+  const toggleExpanded = React.useCallback(
+    (e) => {
+      e.stopPropagation()
+      setExpanded((x) => !x)
+    },
+    [setExpanded],
+  )
+
+  const onClick = React.useCallback((e) => {
+    e.stopPropagation()
+  }, [])
+
+  const path = (prefix || '') + name
+
+  const onDrop = React.useCallback(
+    (files) => {
+      dispatch(FilesAction.Add({ prefix: path, files }))
+    },
+    [dispatch, path],
+  )
+
+  const { getRootProps, isDragActive } = useDropzone({
+    onDrop,
+    noDragEventsBubbling: true,
+  })
+
+  const handle = React.useCallback(
+    (cons) => (e) => {
+      e.stopPropagation()
+      dispatch(cons(path))
+    },
+    [dispatch, path],
+  )
+
+  const action = React.useMemo(
+    () =>
+      ({
+        added: { hint: 'Remove', icon: 'clear', handler: handle(FilesAction.RevertDir) },
+        modified: {
+          hint: 'Revert',
+          icon: 'restore',
+          handler: handle(FilesAction.RevertDir),
+        },
+        deleted: {
+          hint: 'Restore',
+          icon: 'restore',
+          handler: handle(FilesAction.RevertDir),
+        },
+        unchanged: {
+          hint: 'Delete',
+          icon: 'clear',
+          handler: handle(FilesAction.DeleteDir),
+        },
+      }[type]),
+    [type, handle],
+  )
+
+  return (
+    <div
+      {...getRootProps({
+        className: cx(classes.root, classes[type], { [classes.active]: isDragActive }),
+        onClick,
+      })}
+    >
+      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events */}
+      <div onClick={toggleExpanded} className={classes.head} role="button" tabIndex={0}>
+        <M.Icon className={classes.icon}>{expanded ? 'folder_open' : 'folder'}</M.Icon>
+        <div className={classes.name}>{name}</div>
+        <M.IconButton onClick={action.handler} title={action.hint} size="small">
+          <M.Icon fontSize="inherit">{action.icon}</M.Icon>
+        </M.IconButton>
+        <div className={classes.bar} />
+        {!children.length && <div className={classes.empty}>{'<EMPTY DIRECTORY>'}</div>}
```

Comment:
I tried drag'n'drop empty directory, it doesn't show up in the tree.
Tried to remove files from the existing directory and push package, this directory is gone.

How can I reproduce an empty directory in the tree?

Reply:
you can't (without editing the code) -- it's there mostly for the consistency of the model, to cover all possible states (maybe for some hypothetic future use case), but you can't trigger that state from the current ui